addons/web/static/tests/views/list_tests.js

```diff
@@ -11299,6 +11299,67 @@ QUnit.module('Views', {
 
         list.destroy();
     });
+
+    QUnit.test("Auto save: save on closing tab/browser", async function (assert) {
+        assert.expect(1);
+
+        const list = await createView({
+            View: ListView,
+            model: 'foo',
+            data: this.data,
+            arch: `
+                <tree editable="top">
+                    <field name="foo"/>
+                </tree>`,
+            mockRPC(route, { args, method, model }) {
+                if (model === 'foo' && method === 'write') {
+                    assert.deepEqual(args, [
+                        [1],
+                        { foo: 'test' },
+                    ]);
+                }
+                return this._super(...arguments);
+            },
+        });
+
+        await testUtils.dom.click(list.$('.o_field_cell[name="foo"]:first'));
+        await testUtils.fields.editInput(list.$('.o_field_widget[name="foo"]'), "test");
+
+        window.dispatchEvent(new Event("beforeunload"));
+        await testUtils.nextTick();
+
+        list.destroy();
+    });
+
+    QUnit.test("Auto save: save on closing tab/browser (invalid field)", async function (assert) {
+        assert.expect(1);
+
+        const list = await createView({
+            View: ListView,
+            model: 'foo',
+            data: this.data,
+            arch: `
+                <tree editable="top">
+                    <field name="foo" required="1"/>
+                </tree>`,
+            mockRPC(route, { args, method, model }) {
+                if (model === 'foo' && method === 'write') {
+                    assert.step('save'); // should not be called
+                }
+                return this._super(...arguments);
+            },
+        });
+
+        await testUtils.dom.click(list.$('.o_field_cell[name="foo"]:first'));
+        await testUtils.fields.editInput(list.$('.o_field_widget[name="foo"]'), '');
+
+        window.dispatchEvent(new Event("beforeunload"));
+        await testUtils.nextTick();
+
+        assert.verifySteps([], 'should not save because of invalid field');
+
+        list.destroy();
+    });
```

Comment:
there are a lot of tests, which is good, but I think at least two tests are missing: the ones involving onchanges, one for form, one form list. I bet that the latter won't pass.

Reply:
also, I'd keep this commit apart.